pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -0,0 +1,200 @@
+	"os"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/checkpoint"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/fileset"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/fingerprint"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/reader"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+)
+
+var errTooManyActiveFiles = errors.New("number of actively read files exceeds max_concurrent_files")
+
+type Tracker struct {
+	*zap.SugaredLogger
+
+	ReaderFactory reader.Factory
+	maxBatchFiles int
+
+	currentPollFiles  *fileset.Fileset[*reader.Reader]
+	previousPollFiles *fileset.Fileset[*reader.Reader]
+	knownFiles        []*fileset.Fileset[*reader.Metadata]
+}
+
+type trackerOp func(tracker *Tracker)
+
+func New(logger *zap.SugaredLogger, readerFactory reader.Factory, maxBatchFiles int) *Tracker {
+	knownFiles := make([]*fileset.Fileset[*reader.Metadata], 3)
+	for i := 0; i < len(knownFiles); i++ {
+		knownFiles[i] = fileset.New[*reader.Metadata](maxBatchFiles)
+	}
+	return &Tracker{
+		SugaredLogger:     logger.With("component", "fileconsumer"),
+		ReaderFactory:     readerFactory,
+		maxBatchFiles:     maxBatchFiles,
+		currentPollFiles:  fileset.New[*reader.Reader](maxBatchFiles),
+		previousPollFiles: fileset.New[*reader.Reader](maxBatchFiles),
+		knownFiles:        knownFiles,
+	}
+}
+
+func (t *Tracker) PostPoll(persister operator.Persister) {
+	// Any new files that appear should be consumed entirely
+	t.ReaderFactory.FromBeginning = true
+	t.Persist(persister)
+	// rotate at end of every poll()
+	t.rotateFilesets()
+}
+
+func (t *Tracker) OpenFile(path string) {
+	fmt.Println(t.currentPollFiles.Len(), t.previousPollFiles.Len())
+	// if t.currentPollFiles.Len()+t.previousPollFiles.Len() > t.maxBatchFiles*2 {
+	// 	t.Errorw("cannot open file", zap.Error(errTooManyActiveFiles))
+	// 	return
+	// }
+	fp, file := t.makeFingerprint(path)
+	if fp == nil {
+		return
+	}
+
+	// Exclude duplicate paths with the same content. This can happen when files are
+	// being rotated with copy/truncate strategy. (After copy, prior to truncate.)
+	if r := t.currentPollFiles.Match(fp, fileset.Equal); r != nil {
+		// re-add the reader as Match() removes duplicates
+		t.currentPollFiles.Add(r)
+		if err := file.Close(); err != nil {
+			t.Debugw("problem closing file", zap.Error(err))
+		}
+		return
+	}
+
+	r, err := t.newReader(file, fp)
+	if err != nil {
+		t.Errorw("Failed to create reader", zap.Error(err))
+		return
+	}
+
+	t.currentPollFiles.Add(r)
+}
+
+func (t *Tracker) CurrentPollFiles() *fileset.Fileset[*reader.Reader] {
+	return t.currentPollFiles
+}
+
+func (t *Tracker) PreviousPollFiles() *fileset.Fileset[*reader.Reader] {
+	return t.previousPollFiles
+}
+
+func (t *Tracker) Persist(persister operator.Persister) {
```

Comment:
I think persistence is also not a concern of the tracker. We can have a method like `AllMetadata() []*reader.Metadata` which is called instead.